website/docs/index.md

```diff
@@ -53,6 +53,41 @@ ln -s -f ../../scripts/pre-commit .git/hooks/pre-commit
 ```
 
 
+## How to Create a Good UNIX Interface
+
+One of our goals with this project is learning how to write programs and, more importantly,
+the importance of not writing code vs. writing code. 
+
+Here are some 
+- [good rules for writing good unix programs](https://lasr.cs.ucla.edu/ficus-members/geoff/interfaces.html)
+
+A good example can be seen in notions of progress bars. The idea of progress bars has come up twice now, from 
+two different contributors, once in cp, and once in dd.
```

Comment:
This is a great discussion. I will pick on flashrom since some of that affected me and I worked on the original.

Originally the flashrom -V option was a default: it always output. The guy who wrote v1 made it that way, and further used all kinds of cursor addressing. I had to make it optional because it's not possible to flash 1023 systems at the same time and get anything usable out of that output. Oh, but wait! Now i'm flashing 1023 nodes and can't see progress? Now what?

Now what some projects do is the following: pair up a flashrom-like program with a system based on something like expect or similar that can tease the output apart and present 1023 streams of data you can comprehend. The result is special purpose output and a special purpose parser. It's not great. And it's sure not Unix.

With the expvar stuff we have a reasonable way to get information about progress and status, and it does not matter if the program is local or remote, and no matter how many we're running. And we don't have to write special output formats to stderr and write special per-program parsers. That's very powerful. 

I don't think POSIX is actually a positive example. I was around when it was happening. POSIX was an attempt to codify standards based on existing practice and code at the time; some of the explanations for what things meant were after-the-fact attempts to explain the code. Many of the people who did the hard lifting on Unix concepts were not involved in POSIX and it shows. 

The thing I have seen is that there is a tendency to assume that POSIX was a design. In fact it came into the game over 15 years after the real work was done, and it in many ways represented the triumph of committee politics over design. Just read anything Rob ever wrote about how the builtin command worked, for example.

Just because we have stdout and stderr and use them a certain way doesn't mean there are not better ways, and it certainly doesn't mean that what GNU did is anything approaching correct.